config/tools/config_generator.cpp

```diff
@@ -0,0 +1,897 @@
+                const auto copyOptions
+                    = std::filesystem::copy_options::overwrite_existing;
+                std::filesystem::copy(dir_entry, build_dir, copyOptions);
+            }
+        }
+    }
+
+    template<typename T>
+    [[maybe_unused]] auto config_generator::find_value(const std::string key,
+                                                       T& output) -> bool {
+        if(template_options.find(key) != template_options.end()) {
+            if(std::holds_alternative<T>(template_options.at(key))) {
+                output = std::get<T>(template_options.at(key));
+                return true;
+            } else {
+                std::cout << "Warning: Unknown type for " << key
+                          << " template parameter." << std::endl;
+                return false;
+            }
+        } else {
+            std::cout << "Warning: Missing " << key << " template parameter."
+                      << std::endl;
+            return false;
+        }
+    }
+
+    [[maybe_unused]] auto config_generator::generate_configuration_file()
+        -> std::string {
+        // Cumulative return message
+        std::string return_msg = "";
+        std::string output_filename = "tmp.cfg";
+
+        if(!template_file_is_valid) {
+            return_msg += "File provided, " + m_template_config_file
+                        + ", does not exist. Aborting operation. \n";
+            return return_msg;
+        }
+        std::map<std::string, std::string> config_map;
+        load_template(m_template_config_file, config_map);
+        // Deal with all config values in file with "tmpl_"
+        for(auto config_map_it = config_map.begin();
+            config_map_it != config_map.end();
+            config_map_it++) {
+            if(config_map_it->first.find(template_prefix)
+               != std::string::npos) {
+                template_options.emplace(
+                    config_map_it->first,
+                    parse_value(config_map_it->second, false));
+
+            } else {
+                value_t parsed_val = parse_value(config_map_it->second, true);
+                if(std::holds_alternative<size_t>(parsed_val)) {
+                    set_param_to_config_file(config_map_it->first,
+                                             std::get<size_t>(parsed_val));
+                } else if(std::holds_alternative<double>(parsed_val)) {
+                    set_param_to_config_file(config_map_it->first,
+                                             std::get<double>(parsed_val));
+                } else if(std::holds_alternative<std::string>(parsed_val)) {
+                    set_param_to_config_file(
+                        config_map_it->first,
+                        std::get<std::string>(parsed_val));
+                } else {
+                    return_msg += "Warning: Unrecognized type for param, "
+                                + config_map_it->first + ". \n";
+                }
```

Comment:
See comments above. (I'm not sure if this would actually be cleaner with a type-matching visitor, but it does look like the `else` case here is also unreachable.)

Reply:
Just seeing this now after exploring using visit for a while. I agree with your assessment here. I don't think using a type-matching visit is any better. I will leave both as they are but remove the unreachable else statement. I figured its good practice to always have an else just in case, say, the parse_value code changes at some point. But i will remove them for now.

DONE